Configurable s3 storage class

inbarpatashnik opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

#### What this PR does
Added support in configurable s3 storage class.

#### Which issue(s) this PR fixes or relates to

Fixes #3438 

#### Checklist

- [X] Tests updated
- [X] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/mimir?pullRequest=4300) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/mimir?pullRequest=4300) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/mimir?pullRequest=4300) it.</sub>

inbarpatashnik commented:
I haven't tried to run it against a real s3.
So I would really appreciate if you could also check it after reviewing  :)

charleskorn commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## main / unreleased
+* [FEATURE] Add capability to configure the used S3 Storage Class [#3438] (https://github.com/grafana/mimir/pull/4300) (@inbarpatashnik)
```
Could you please move this underneath the 'Grafana Mimir' heading?

charleskorn replied:
Please also follow the same format as for other changelog entries:

```suggestion
* [FEATURE] Add capability to configure the S3 storage class. #3438
```

charleskorn commented on pkg/storage/bucket/s3/bucket_client.go:
```diff
@@ -12,6 +12,11 @@ import (
 
+const (
+	// Applied to PUT operations to denote the desired storage class for S3 Objects
+	awsStorageClassHeader = "X-Amz-storage-Class"
```
[nit] Could you please make the casing consistent here?
```suggestion
	awsStorageClassHeader = "X-Amz-Storage-Class"
```

charleskorn commented on pkg/storage/bucket/s3/config.go:
```diff
@@ -32,13 +32,26 @@ const (
 var (
 	supportedSignatureVersions     = []string{SignatureVersionV4, SignatureVersionV2}
 	supportedSSETypes              = []string{SSEKMS, SSES3}
+	supportedStorageClasses        = []string{StorageClassGlacier, StorageClassDeepArchive, StorageClassGlacierInstantRetrieval, StorageClassIntelligentTiering, StorageClassOneZoneInfrequentAccess, StorageClassOutposts, StorageClassReducedRedundancy, StorageClassStandard, StorageClassStandardInfrequentAccess}
```
Do we need to define these constants ourselves? It looks like the S3 SDK exposes them with [`ObjectStorageClass_Values()`](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#ObjectStorageClass_Values) and the `ObjectStorageClass...` constants.

charleskorn commented on pkg/storage/bucket/s3/config.go:
```diff
@@ -32,13 +32,26 @@ const (
+	supportedStorageClasses        = []string{StorageClassGlacier, StorageClassDeepArchive, StorageClassGlacierInstantRetrieval, StorageClassIntelligentTiering, StorageClassOneZoneInfrequentAccess, StorageClassOutposts, StorageClassReducedRedundancy, StorageClassStandard, StorageClassStandardInfrequentAccess}
 	errUnsupportedSignatureVersion = fmt.Errorf("unsupported signature version (supported values: %s)", strings.Join(supportedSignatureVersions, ", "))
 	errUnsupportedSSEType          = errors.New("unsupported S3 SSE type")
+	errUnsupportedStorageClass     = errors.New("unsupported S3 storage class")
```
Would be good to include the list of supported values in the error, similar to what's done for `errUnsupportedSignatureVersion` above.

charleskorn commented on pkg/storage/bucket/s3/bucket_client.go:
```diff
@@ -39,13 +44,14 @@ func newS3Config(cfg Config) (s3.Config, error) {
+		AccessKey:       cfg.AccessKeyID,
+		SecretKey:       cfg.SecretAccessKey.String(),
+		Insecure:        cfg.Insecure,
+		PutUserMetadata: map[string]string{awsStorageClassHeader: cfg.StorageClass},
```
Might be being overly cautious here: it's probably safest for the default value of `cfg.StorageClass` to be an empty string, and to not set `PutUserMetadata` at all if `StorageClass` is not set - some S3-compatible storage services may not support `STANDARD`.

inbarpatashnik replied:
Hi! thank you so much for the review!
It sounds like a very good point - Ive tried to edit so the default value of the storage class in the config.go will be an empty string, but it seemed to fail some tests because the object storage class "" doesn't exist.
Maybe I should edit it in this file to be an empty string?
I fixed the pr according to all the comments, but I still think how to fix so solve this one.

inbarpatashnik replied:
And also I'm thinking - maybe its better if it will have an actual default value that will be supported by default (If it exists?)

charleskorn replied:
Rather than setting it to an empty string, I would simply not set it at all (which is what we were doing before).

charleskorn reviewed (commented):
Thanks for the PR @inbarpatashnik!

charleskorn commented on go.mod:
```diff
@@ -88,7 +89,6 @@ require (
 	github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751 // indirect
 	github.com/armon/go-metrics v0.4.0 // indirect
 	github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d // indirect
-	github.com/aws/aws-sdk-go v1.44.187 // indirect
```
Is the version bump needed for this change? It seems unrelated (but I might be missing something). If the version bump is not required, could you please change this PR to use the previous version?

inbarpatashnik replied:
the thing is when I ran the _go mod vendor_ this was the output to the file

inbarpatashnik replied:
I ran it because of the new import

inbarpatashnik replied:
> Is the version bump needed for this change? It seems unrelated (but I might be missing something). If the version bump is not required, could you please change this PR to use the previous version?

The thing is when I ran the _go mod vendor_ because of the new import, this was the output to the file

charleskorn replied:
Have you tried editing `go.mod` to use the same version as before?

charleskorn commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@
 
 ### Grafana Mimir
 
+* [FEATURE] Add capability to configure the S3 storage class. #3438
+
```
[nit] Blank line not required.

```suggestion
* [FEATURE] Add capability to configure the S3 storage class. #3438
```

charleskorn reviewed (commented):
I think we can still make this change safer - see [this comment thread](https://github.com/grafana/mimir/pull/4300/files#r1125764753). Please also rebase to resolve the conflicts.

charleskorn commented on integration/alertmanager_test.go:
Is setting `StorageClass` to `STANDARD` required? Couldn't we continue using the default value, like before?

charleskorn replied:
(also applies to other integration test files below)

inbarpatashnik replied:
It needs to be some type of storage class, because otherwise it will throw the invalid storage class.

inbarpatashnik replied:
Oh I see you said we need to allow an empty string as a valid storage class.
That would fix it :)

charleskorn commented on pkg/storage/bucket/s3/config.go:
```diff
@@ -113,6 +118,9 @@ func (cfg *Config) Validate() error {
 			return errInvalidEndpointPrefix
 		}
 	}
+	if !util.StringsContain(supportedStorageClasses, cfg.StorageClass) {
```
We'd also need to allow an empty string.

charleskorn commented on pkg/storage/bucket/s3/config.go:
```diff
@@ -79,6 +83,7 @@ type Config struct {
 	AccessKeyID      string         `yaml:"access_key_id"`
 	Insecure         bool           `yaml:"insecure" category:"advanced"`
 	SignatureVersion string         `yaml:"signature_version" category:"advanced"`
+	StorageClass     string         `yaml:"storage_class"`
```
Please also add a CLI flag for this option in `RegisterFlagsWithPrefix` below.

charleskorn commented:
@inbarpatashnik I'm not sure if you intended to ask for a re-review from me - there have been no changes pushed to this PR since [my last review](https://github.com/grafana/mimir/pull/4300#pullrequestreview-1349740031).

inbarpatashnik commented:
> @inbarpatashnik I'm not sure if you intended to ask for a re-review from me - there have been no changes pushed to this PR since [my last review](https://github.com/grafana/mimir/pull/4300#pullrequestreview-1349740031).

I added in config.go :
- changed the condition in line 121 so an exception won't be raised if the storage class is empty string
- added the storage class cli flag

And I removed the changes I've made in the test where I defined a storage class when it isn't really necessary when there is a support in the empty string

inbarpatashnik commented:
> > @inbarpatashnik I'm not sure if you intended to ask for a re-review from me - there have been no changes pushed to this PR since [my last review](https://github.com/grafana/mimir/pull/4300#pullrequestreview-1349740031).
> 
> I added in config.go :
> - changed the condition in line 121 so an exception won't be raised if the storage class is empty string
> - added the storage class cli flag
> 
> And I removed the changes I've made in the test where I defined a storage class when it isn't really necessary when there is a support in the empty string

I pushed it to my fork's main branch, it didn't sync here for some reason.

charleskorn commented:
> I pushed it to my fork's main branch, it didn't sync here for some reason.

This PR is pulling from the `fixture/conflicts` branch in your fork. You'll need to either change the source branch for this PR, or push those changes to the `fixture/conflicts` branch.

inbarpatashnik commented:
> > I pushed it to my fork's main branch, it didn't sync here for some reason.
> 
> This PR is pulling from the `fixture/conflicts` branch in your fork. You'll need to either change the source branch for this PR, or push those changes to the `fixture/conflicts` branch.

@charleskorn I don't know how it happend or how to change it back to my main branch, but I merged the main to the fixture/conflicts :)

charleskorn commented on pkg/storage/bucket/s3/config.go:
```diff
@@ -113,6 +119,9 @@ func (cfg *Config) Validate() error {
 			return errInvalidEndpointPrefix
 		}
 	}
+	if !util.StringsContain(supportedStorageClasses, cfg.StorageClass) && (cfg.StorageClass != "") {
```
[nit] Extra parentheses are not required 

```suggestion
	if !util.StringsContain(supportedStorageClasses, cfg.StorageClass) && cfg.StorageClass != "" {
```

charleskorn commented on integration/alertmanager_test.go:
Are all of the `StorageClass: "STANDARD"` changes to the integration tests required now that the default empty value is permitted? If not, then please remove them, it would be good to confirm that the tests continue to behave as they did before with the default config.

inbarpatashnik replied:
I think I returned it to the original form in all the relevant tests :)

charleskorn reviewed (commented):
This is looking good to me @inbarpatashnik. Only thing remaining to address is [this comment](https://github.com/grafana/mimir/pull/4300/files#r1125764753).

charleskorn reviewed (changes requested):
https://github.com/grafana/mimir/pull/4300/commits/34a746f9e898264af64f6b2e7d9b79e5a8456249 went too far 😄 

[This comment](https://github.com/grafana/mimir/pull/4300/files#r1125764753) was a suggestion to only set the `X-Amz-Storage-Class` value if `cfg.StorageClass` was not an empty string. 

So, I'm imagining something like this:

```go
func newS3Config(cfg Config) (s3.Config, error) {
	sseCfg, err := cfg.SSE.BuildThanosConfig()
	if err != nil {
		return s3.Config{}, err
	}

	putUserMetadata := map[string]string{}

	if cfg.StorageClass != "" {
		putUserMetadata[awsStorageClassHeader] = cfg.StorageClass
	}

	return s3.Config{
		Bucket:          cfg.BucketName,
		Endpoint:        cfg.Endpoint,
		Region:          cfg.Region,
		AccessKey:       cfg.AccessKeyID,
		SecretKey:       cfg.SecretAccessKey.String(),
		Insecure:        cfg.Insecure,
		PutUserMetadata: putUserMetadata,
		SSEConfig:       sseCfg,
		HTTPConfig: s3.HTTPConfig{
			IdleConnTimeout:       model.Duration(cfg.HTTP.IdleConnTimeout),
			ResponseHeaderTimeout: model.Duration(cfg.HTTP.ResponseHeaderTimeout),
			InsecureSkipVerify:    cfg.HTTP.InsecureSkipVerify,
			TLSHandshakeTimeout:   model.Duration(cfg.HTTP.TLSHandshakeTimeout),
			ExpectContinueTimeout: model.Duration(cfg.HTTP.ExpectContinueTimeout),
			MaxIdleConns:          cfg.HTTP.MaxIdleConns,
			MaxIdleConnsPerHost:   cfg.HTTP.MaxIdleConnsPerHost,
			MaxConnsPerHost:       cfg.HTTP.MaxConnsPerHost,
			Transport:             cfg.HTTP.Transport,
		},
		// Enforce signature version 2 if CLI flag is set
		SignatureV2: cfg.SignatureVersion == SignatureVersionV2,
	}, nil
}
```

inbarpatashnik commented:
ohhh felt like it.
I understand, now it doesn't feel weird 🤣

charleskorn commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 ### Grafana Mimir
 
+* [FEATURE] Add capability to configure the S3 storage class. #3438
```
```suggestion
* [FEATURE] Added `-<prefix>.s3.storage-class` flag to configure the S3 storage class for objects written to S3 buckets. #3438
```

charleskorn replied:
Hope you don't mind @inbarpatashnik, going to commit this minor edit so we can get this merged :)

charleskorn reviewed (approved):
Thanks again @inbarpatashnik!

inbarpatashnik commented:
> Thanks again @inbarpatashnik!

Thank you!!🤗

56quarters commented on pkg/storage/bucket/s3/config_test.go:
```diff
@@ -7,12 +7,14 @@ package s3
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+
+	"net/http"
```
Can you move these with the other stdlib imports?

56quarters commented on pkg/storage/bucket/s3/config.go:
```diff
@@ -98,6 +103,7 @@ func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.StringVar(&cfg.Endpoint, prefix+"s3.endpoint", "", "The S3 bucket endpoint. It could be an AWS S3 endpoint listed at https://docs.aws.amazon.com/general/latest/gr/s3.html or the address of an S3-compatible service in hostname:port format.")
 	f.BoolVar(&cfg.Insecure, prefix+"s3.insecure", false, "If enabled, use http:// for the S3 endpoint instead of https://. This could be useful in local dev/test environments while using an S3-compatible backend storage, like Minio.")
 	f.StringVar(&cfg.SignatureVersion, prefix+"s3.signature-version", SignatureVersionV4, fmt.Sprintf("The signature version to use for authenticating against S3. Supported values are: %s.", strings.Join(supportedSignatureVersions, ", ")))
+	f.StringVar(&cfg.StorageClass, prefix+"s3.storage-class", "", "The S3 storage class to use. Details can be found at https://aws.amazon.com/s3/storage-classes/. Supported values are: "+strings.Join(supportedStorageClasses, ", "))
```
Should we mention the that if this is unset, the default for s3 will be used?

56quarters commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 ### Grafana Mimir
 
+* [FEATURE] Added `-<prefix>.s3.storage-class` flag to configure the S3 storage class for objects written to S3 buckets. #3438
```
Can you move this next to the other `[FEATURE]` entries.

56quarters reviewed (approved):
LGTM with some minor tweaks. Thanks!

inbarpatashnik commented on pkg/storage/bucket/s3/config.go:
```diff
@@ -9,15 +9,17 @@ import (
 	"github.com/pkg/errors"
 	"github.com/thanos-io/objstore/providers/s3"
 
+	"net/http"
```
@56quarters i tried to reorganize them, but it seems to fail in the lint :(
